debian/apparmor/ubuntu_pro_esm_cache.jinja2

```diff
@@ -0,0 +1,289 @@
+
+  /usr/bin/cloud-id Cx -> cloud_id,
+  # in bionic, it's /bin/ps, so let's match both
+  /{,usr/}bin/ps Cx -> ps,
+  /usr/bin/systemd-detect-virt Px -> ubuntu_pro_esm_cache_systemd_detect_virt,
+  /usr/bin/dpkg Cx -> dpkg,
+  /usr/bin/ubuntu-distro-info Cx -> ubuntu_distro_info,
+  /usr/lib/apt/methods/gpgv Cx -> apt_methods_gpgv,
+  /usr/lib/apt/methods/http Cx -> apt_methods,
+  /usr/lib/apt/methods/https Cx -> apt_methods,
+  /usr/lib/apt/methods/store Cx -> apt_methods,
+  # when there is no status.json cached, esm-cache.service will invoke "snap status"
+  /usr/bin/snap PUx,
+
+  /usr/share/dpkg/** r,
+  /usr/share/keyrings/* r,
+
+  /var/cache/apt/** rw,
+
+  /var/lib/apt/** r,
+  /var/lib/dpkg/** r,
+  /var/lib/ubuntu-advantage/** rwk,
+
+  /var/log/ubuntu-advantage.log rw,
+
+  @{PROC}/@{pid}/fd/ r,
+  @{PROC}/1/cgroup r,
+  @{PROC}/version_signature r,
+  @{PROC}/@{pid}/mountinfo r,
+  @{PROC}/@{pid}/status r,
+  @{PROC}/@{pid}/stat r,
+  @{PROC}/sys/kernel/osrelease r,
+
+{% if ubuntu_codename in ["bionic", "xenial"] %}
+  # see https://bugs.python.org/issue40501
+  /sbin/ldconfig rix,
+  /sbin/ldconfig.real rix,
+  @{PROC}/@{pid}/mounts r,
+  /usr/bin/@{multiarch}-gcc-* rix,
+  /usr/bin/@{multiarch}-ld.bfd rix,
+  /usr/lib/gcc/@{multiarch}/*/collect2 rix,
+  /usr/bin/@{multiarch}-objdump rix,
+{% endif %}
+
+  profile ps flags=(attach_disconnected) {
+    include <abstractions/base>
+    include <abstractions/nameservice>
+
+    capability sys_ptrace,
+    ptrace read,
+
+    # in bionic, it's /bin/ps, so let's match both
+    /{,usr/}bin/ps mrix,
+
+    /dev/tty r,
+
+    @{PROC}/ r,
+    @{PROC}/@{pid}/** r,
+    @{PROC}/uptime r,
+    @{PROC}/sys/kernel/** r,
+  }
+
+  profile cloud_id flags=(attach_disconnected) {
+    include <abstractions/base>
+    include <abstractions/nameservice>
+    include <abstractions/python>
+
+    ptrace read peer=unconfined,
+
+    /etc/cloud/** r,
+    /etc/apt/** r,
+    /etc/apport/** r,
+    /etc/ssl/openssl.cnf r,
+
+    @{PROC}/@{pid}/fd/ r,
+    @{PROC}/cmdline r,
+    @{PROC}/1/environ r,
+    @{PROC}/1/cmdline r,
+    @{PROC}/@{pid}/status r,
+
+    /run/cloud-init/** r,
+
+    /usr/bin/ r,
+    /usr/bin/cloud-id r,
+    /usr/bin/python3.{1,}[0-9] mrix,
+    /usr/bin/uname mrix,
+
+    /usr/share/dpkg/** r,
+
+    # workarounds for
+    # https://gitlab.com/apparmor/apparmor/-/issues/346
+    /usr/bin/systemctl Px -> ubuntu_pro_esm_cache_systemctl,
+    /usr/bin/systemd-detect-virt Px -> ubuntu_pro_esm_cache_systemd_detect_virt,
+
+    /var/lib/cloud/** r,
+
+{% if ubuntu_codename in ["bionic", "xenial"] %}
+    # see https://bugs.python.org/issue40501
+    /sbin/ldconfig rix,
+    /sbin/ldconfig.real rix,
+    @{PROC}/@{pid}/mounts r,
+    /usr/bin/@{multiarch}-gcc-* rix,
+    /usr/bin/@{multiarch}-ld.bfd rix,
+    /usr/lib/gcc/@{multiarch}/*/collect2 rix,
+    /usr/bin/@{multiarch}-objdump rix,
+
+    /etc/lsb-release r,
+    @{PROC}/cmdline r,
+    /bin/dash mrix,
```

Comment:
could it change based on the system configs?

Reply:
ah, just saw Seth's comment that this is a test framework, so I guess not

Reply:
What do you mean? The part Seth commented on (if I understood it correctly) was the python code in features/environment.py, that is a test framework. But this profile is for the "real thing": ubuntu-pro-client (a deb) installed on ubuntu releases since xenial.

Reply:
my bad, I misunderstood. I guess the original comment is still valid then, could another shell be executed instead of dash?

Reply:
The exact sequence of calls when I run `cloud-id` on bionic is:
```
TIME(s) PCOMM            PID    PPID   RET ARGS
6.063   cloud-id         3096   3080     0 /usr/bin/cloud-id
6.166   ldconfig         3097   3096     0 /sbin/ldconfig -p
6.166   ldconfig.real    3097   3096     0 /sbin/ldconfig.real -p
6.177   ldconfig         3098   3096     0 /sbin/ldconfig -p
6.177   ldconfig.real    3098   3096     0 /sbin/ldconfig.real -p
6.201   sh               3099   3096     0 /bin/sh -c uname -p 2> /dev/null
6.202   uname            3100   3099     0 /bin/uname -p
6.242   systemd-detect-  3101   3096     0 /usr/bin/systemd-detect-virt --quiet --container
```
The reason that ended up calling `/bin/dash` is because our `/bin/sh` is a symlink to `/bin/dash`. I have never seen it be anything else.

And the reason `/bin/sh` was called is because it's the shebang of `/sbin/ldconfig`:
```terminal
root@b-pro:~# cat /sbin/ldconfig
#!/bin/sh

if  test $# = 0                                                 \
    && test x"$LDCONFIG_NOTRIGGER" = x                          \
 && test x"$DPKG_MAINTSCRIPT_PACKAGE" != x                      \
 && dpkg-trigger --check-supported 2>/dev/null
then
        if dpkg-trigger --no-await ldconfig; then
                if test x"$LDCONFIG_TRIGGER_DEBUG" != x; then
                        echo "ldconfig: wrapper deferring update (trigger activated)"
                fi
                exit 0
        fi
fi

exec /sbin/ldconfig.real "$@"
```
I don't think another shell would be called, but is there a more general abstraction to allow "all"  shells that I should perhaps consider?

Reply:
sorry I missed this notification. yeah, you're right, I don't think any other shell would be called. thanks for checking and for the detailed explanation.